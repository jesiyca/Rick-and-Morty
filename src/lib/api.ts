import { Character, CharactersResponse } from '@/types/character';

interface FetchCharactersParams {
    page?: number;
    name?: string;
    species?: string;
    status?: string;
    type?: string;
    gender?: string;
  }

export async function fetchCharacter(id: number): Promise<Character> {
    const res = await fetch(`https://rickandmortyapi.com/api/character/${id}`);
    if (!res.ok) throw new Error(`Failed to fetch character with id = ${id} with error = ${res.statusText}`);
    return res.json();
  }

export async function fetchCharacters(params: FetchCharactersParams = {}): Promise<CharactersResponse> {
    const query = new URLSearchParams();

    for (const [key, value] of Object.entries(params)) {
      if (value != null && value !== '') {
        query.append(key, String(value));
      }
    }
  
    const res = await fetch(`https://rickandmortyapi.com/api/character/?${query}`);
    if (!res.ok) {
      throw new Error(`Failed to fetch characters for ${query} with error = ${res.statusText}`);
    }
  
    return res.json();
  }