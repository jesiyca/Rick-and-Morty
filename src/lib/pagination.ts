/**
 * Generates a list of pages for pagination UI, with ellipsis (-1) placeholders.
 */
export function getPageRange(currentPage: number, totalPages: number, visiblePages = 5): number[] {
    const pages: number[] = [];
  
    if (totalPages <= visiblePages) {
      return Array.from({ length: totalPages }, (_, i) => i + 1);
    }
  
    const showLeftEllipsis = currentPage > 3;
    const showRightEllipsis = currentPage < totalPages - 2;
  
    pages.push(1);
  
    if (showLeftEllipsis) {
      pages.push(-1); // Left ellipsis
    }
  
    const start = Math.max(2, currentPage - 1);
    const end = Math.min(totalPages - 1, currentPage + 1);
  
    for (let i = start; i <= end; i++) {
      pages.push(i);
    }
  
    if (showRightEllipsis) {
      pages.push(-1); // Right ellipsis
    }
  
    pages.push(totalPages);
  
    return pages;
  }