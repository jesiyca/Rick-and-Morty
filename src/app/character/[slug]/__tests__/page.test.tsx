import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import CharacterPage from '../page';
import { useParams, useRouter } from 'next/navigation';
import { fetchCharacter } from '@/lib/api';

jest.mock('next/navigation', () => ({
  useParams: jest.fn(),
  useRouter: jest.fn(),
}));

jest.mock('@/lib/api', () => ({
  fetchCharacter: jest.fn(),
}));

const mockBack = jest.fn();

const mockCharacter = {
  id: 1,
  name: 'Rick Sanchez',
  status: 'Alive',
  origin: { name: 'Earth' },
  location: { name: 'Citadel of Ricks' },
  image: 'Test Image',
  created: '2023-01-01T00:00:00.000Z',
};

describe('CharacterPage', () => {
  beforeEach(() => {
    (useParams as jest.Mock).mockReturnValue({ slug: '1' });
    (useRouter as jest.Mock).mockReturnValue({ back: mockBack });
    (fetchCharacter as jest.Mock).mockResolvedValue(mockCharacter);
  });

  afterEach(jest.clearAllMocks);

    it('matches snapshot', () => {
      const { container } = render(<CharacterPage />);
      expect(container).toMatchSnapshot();
    });

  it('renders loading initially and then character info', async () => {
    render(<CharacterPage />);
    expect(screen.getByText('Loading character data...')).toBeInTheDocument();

    expect(await screen.findByText('Rick Sanchez')).toBeInTheDocument();
    expect(screen.getByText('Status: Alive')).toBeInTheDocument();
    expect(screen.getByText('Origin: Earth')).toBeInTheDocument();
    expect(screen.getByText('Location: Citadel of Ricks')).toBeInTheDocument();
    expect(screen.getByText('Created: January 1, 2023')).toBeInTheDocument();
  });

  it('calls router.back() when back button is clicked', async () => {
    render(<CharacterPage />);
    await screen.findByText('Rick Sanchez'); // wait until data is loaded

    fireEvent.click(screen.getByRole('button', { name: "Back" }));
    expect(mockBack).toHaveBeenCalled();
  });
});
