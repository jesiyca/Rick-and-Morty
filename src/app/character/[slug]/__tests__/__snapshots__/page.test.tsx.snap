// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterPage matches snapshot 1`] = `
<div>
  <div
    class="grid grid-rows-[20px_1fr_20px] items-center justify-items-center min-h-screen p-8 pb-20 gap-16 sm:p-20 font-[family-name:var(--font-geist-sans)]"
  >
    <main
      class="flex flex-col gap-4 row-start-2 items-center sm:items-start font-normal"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3"
        data-slot="button"
      >
        Back
      </button>
      <p>
        Loading character data...
      </p>
    </main>
  </div>
</div>
`;
